lib/synapse/service_watcher/etcd.rb

```diff
@@ -0,0 +1,166 @@
+        unless @discovery['hosts']
+      raise ArgumentError, "invalid etcd path for service #{@name}" \
+        unless @discovery['path']
+    end
+
+    # helper method that ensures that the discovery path exists
+    def create(path)
+      log.debug "synapse: creating etcd path: #{path}"
+      @etcd.create(path, dir: true)
+    end
+
+    def each_node(node)
+      begin
+        host, port, name = deserialize_service_instance(node.value)
+      rescue StandardError => e
+        log.error "synapse: invalid data in etcd node #{node.inspect} at #{@discovery['path']}: #{e} DATA #{node.value}"
+        nil 
+     else
+        server_port = @server_port_override ? @server_port_override : port
+
+        # find the numberic id in the node name; used for leader elections if enabled
+        numeric_id = node.key.split('/').last
+        numeric_id = NUMBERS_RE =~ numeric_id ? numeric_id.to_i : nil
+
+        log.warn "synapse: discovered backend #{name} at #{host}:#{server_port} for service #{@name}"
+        { 'name' => name, 'host' => host, 'port' => server_port, 'id' => numeric_id}
+      end
+    end
+
+    def each_dir(d)
+      new_backends = []
+      d.children.each do |node|
+        if node.directory?
+          new_backends << each_dir(@etcd.get(node.key))
+        else
+          backend = each_node(node)
+          if backend
+            new_backends << backend
+          end
+        end
+      end
+      new_backends.flatten
+    end
+
+    # find the current backends at the discovery path; sets @backends
+    def discover
+      log.info "synapse: discovering backends for service #{@name}"
+
+      d = nil
+      begin
+        d = @etcd.get(@discovery['path'])
+      rescue Etcd::KeyNotFound
+        create(@discovery['path'])
+        d = @etcd.get(@discovery['path'])
+      end
+
+      new_backends = []
+      if d.directory?
+        new_backends = each_dir(d)
+      else
+        log.warn "synapse: path #{@discovery['path']} is not a directory"
+      end
+
+      if new_backends.empty?
+        if @default_servers.empty?
+          log.warn "synapse: no backends and no default servers for service #{@name}; using previous backends: #{@backends.inspect}"
+          false
+        else
+          log.warn "synapse: no backends for service #{@name}; using default servers: #{@default_servers.inspect}"
+          @backends = @default_servers
+          true
+        end
+      else
+        if @backends != new_backends
+          log.info "synapse: discovered #{new_backends.length} backends (including new) for service #{@name}"
+          @backends = new_backends
+          true
+        else
+          log.info "synapse: discovered #{new_backends.length} backends for service #{@name}"
+          false
+        end
+      end
+    end
+
+    def watch
+      while !@should_exit
+        begin
+          @etcd.watch(@discovery['path'], :timeout => 60, :recursive => true)
```

Comment:
Alright so while testing I came across this gem. If my understanding is correct etcd makes us not only do our own heartbeats from nerve (hella writes), but we also end up basically constantly polling in this function because there is no capability to filter watch events?

Basically without https://github.com/coreos/etcd/issues/633 or https://github.com/coreos/etcd/issues/174 closed SmartStack scalability on etcd will be limited to ... not very much.

If that understanding is correct can we implement get children watches internally, where we do a lightweight "see if the list of children changed" operation and if that is different we actually go through loading all the data (aka run discover). This way we're at least not constantly pulling all of the etcd state?